Add asm and LLVM IR output cmdline options

JohanEngelen opened the pull request:
To help development, add the "--output-ll" cmdline option that saves the LLVM IR to a file. The same as LDC's --output-ll option.

JohanEngelen commented:
How about I split `backend.emitObject()` into two functions (split right before outputting the objfile)? That way, SDC can output asm, LLVM IR, and obj independently.

deadalnix commented:
Yes that sounds like the right path forward. Emitting asm is also very desirable.

JohanEngelen commented:
@deadalnix What do you think after this 2nd commit?

deadalnix commented on sdc/src/sdc/main.d:
```diff
@@ -29,13 +29,16 @@ int main(string[] args) {
+		"c",         "Stop before linking", &dontLink,
+		"o",         "Output file",         &outputFile,
+		"output-ll", "Output LLVM IR",      &outputIR,
+		"output-s",  "Output assembly",     &outputAsm
```
-S

JohanEngelen commented:
I've changed the flags to match clang's (from what I understood from documentation), and added clang's LLVM bitcode output option.
`-S`  will output assembly (and stop)
`-S --emit-llvm`   will output LLVM assembly
`-c`   will output an object file (and stop)
`-c --emit-llvm`  will output LLVM bitcode

deadalnix commented on libd-llvm/src/d/llvm/backend.d:
```diff
@@ -119,22 +117,13 @@ public:
-		
 		char* errorPtr;
-		auto linkError = LLVMTargetMachineEmitToFile(targetMachine, dmodule, objFile.toStringz(), LLVMCodeGenFileType.Object, &errorPtr);
+		auto linkError = LLVMTargetMachineEmitToFile(targetMachine, pass.dmodule, objFile.toStringz(), LLVMCodeGenFileType.Object, &errorPtr);
```
Is pass.dmodule necessary here ?

JohanEngelen replied:
How else would LLVM know what to write to the file? ;-)

deadalnix replied:
I meant, as opposed to use dmodule directly, but pass is not aliasthised here, so indeed.

deadalnix commented:
Ok some nits, but this is lookign really good, I like it.
First, could you make sure that failure cause the compiler to return with a non 0 exit code (important for build systems) ? I think the best way to achieve this is to throw a CompileError.
Second, that would be nice to have some kind of test for these behaviors.

JohanEngelen commented:
A `CompileError` seems off (for example, what would be the `Location` of the error?). It's probably better to add another Exception to the hierarchy? That'd be more work, kind of unrelated to this... For now, I could copy what is done currently for outputing obj files: `assert(0, "Fail to link ! Exiting...");`. Looks like SDC does not quit with non-0 code for obj file errors now. Perhaps a separate PR then? 

(my current PRs are to get familiar with SDC development, I'd rather not linger too much on this and move on to more interesting pieces :P)

JohanEngelen commented:
I agree about missing tests. Perhaps after this lands and SDC can output asm and LLVM IR, a new type of test can be added that checks if the outputted asm/LLVM is correct. (for example, a check for symbol name mangling, debug info, LLVM attributes, etc.)

deadalnix commented:
It is not a bug that it emits a .o file, but I guess it should do it like gcc/clang do it: in /tmp with a randomized name and blast it away when linked, unless the object file is actually needed.

deadalnix commented:
I'm good with assert(0) here. That's not super duper nice, but will do for now.

JohanEngelen commented:
You want me to add that to runner.d?

deadalnix commented:
I'm not sure what is the best way forward here. But the problem with code that do not run, is that we won't notice when we break it. So that would be great to have some file compiled with the various flags and at least to check the output is what is expected.

JohanEngelen commented:
@deadalnix  OK, I am reworking the tests to be run by LLVM's lit. This gives much more flexibility in testing. The idea is to enable SDC to output to stdout, and pipe that to a checker tool. The end result will have files that look something like:

```
// RUN: %sdc %s -S --emit-llvm -o - | %FileCheck %s

int main() {
// CHECK: _D8test00004mainFMZi
	return 42+1;
// CHECK: 43
}
```

Seems like a very powerful way to test the compiler. (clang is using this method)

TODO: 
- add `// RUN:` lines to test suite files (in progress)
- add stdout output option  (`sdc test.d -S --emit-llvm -o -` ?)
- make LLVM's 'lit' python script work for everybody and Travis

JohanEngelen commented:
@deadalnix Would be great to have your "OK" on the conversion to `lit` as test tool, before I spend more time on it. This would replace `runner.d`.  So far, I like it a lot.

JohanEngelen commented:
I ran into issue #170 while working on this. The work-around for now is to add `module ...` to a few test files to hardcode the module name.

JohanEngelen commented:
I added `lit` in the tests folder, but I think this thing should be put in a `utils` folder instead, but much better would be to not have this in the repo at all. `lit` is available in PyPi: https://pypi.python.org/pypi/lit , but I don't know how easy it is for people/Travis to install it.

Unfortunately, again, I cannot read Travis's cmdline log (it does not load in my browser), so can't see where it errors :(.

deadalnix commented:
Hi,

First, yeah I thing adding a test suite based on llvm-lit is a good idea. I actually was considering adding that at some point. That being said, I don't think converting everything to llvm-lit is especially a good idea. It is also undesirable to convert everything to lit just to get this in. This only needs to run the compiler with the various command line switch on a random file.

I'd suggest having a llvm subfolder in the test folder and put lit tests in there. @schancel how would we go about adding Filecheck and llvm-lit to travis ?

deadalnix commented on sdc/src/sdc/sdc.d:
```diff
@@ -75,14 +75,29 @@ final class SDC {
-	
-	void codeGen(string objFile) {
+
+	void runLLVMPasses() {
```
This expose implementation of the backend and add sequential coupling.

JohanEngelen replied:
I can rename `runLLVMPasses` to `codeGen`. I think it is clearest if main.d tells the backend to compile, codegen, and output stuff (instead of have codegen output a file).
Tell me how you want it, I have no preference.

deadalnix replied:
I don't think this should be exposed at all. It needs to run when generation any kind of file, so functions that generate other files, so these function should run the passes on a per needed basis.

deadalnix commented on tests/lit.py:
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env python
```
Is it really needed to add the full lit program in there ? It is already in llvm.

JohanEngelen replied:
lit is not included in macport's LLVM install. And I wasn't sure about where you want to get it from (PyPi, developer install, LLVM source...). My preference is also to not have this in SDC's repo.

deadalnix commented on tests/test0000.d:
```diff
@@ -1,3 +1,5 @@
+// RUN: %sdc %s -o %t && sh -c '%t; [ $? = 42 ]'
```
I don't think this is the right way forward. We should use this to check the codegen for LLVM, but I'd like to see all this test suite ran by runner.d using libd as a library rather than running sdc again and again.

JohanEngelen replied:
An advantage of using lit is that you can easily check for stdout too (for the tests that are not supposed to compile). Imho, lit is basically an improved runner.d that SDC does not need to maintain.
At the moment, runner.d is running SDC again and again. Were you planning on changing runner.d?

deadalnix replied:
Well, @UplinkCoder had a good first shot at it, but never finished it. I still think that it was the right way forward and hope this can be nailed at some point.

Also I sent you a slack invite so we can talk on a more direct way.

JohanEngelen commented:
(rebased for cleaner history)

deadalnix commented on tests/runner.d:
```diff
@@ -186,6 +218,12 @@ int main(string[] args)
         return regressed ? -1 : 0;
     }
 
+    // Run temporary tests, remove when proper assembly output tests are in place.
```
Prefix with TODO if you know what to do, FIXME if this si know to be broken but ok for now, or XXX if this is wonky but the path forward is not known.

deadalnix commented on libd-llvm/src/d/llvm/backend.d:
```diff
@@ -119,22 +117,15 @@ public:
-		LLVMRunPassManager(pm, dmodule);
-		
-		// Dump module for debug purpose.
-		// LLVMDumpModule(dmodule);
```
Can we keep these around ? They are pretty useful.

deadalnix replied:
ping ?

JohanEngelen replied:
forgot about this. 
(I deleted it, as I thought one could now easily do this with --emit-llvm without rebuilding sdc, but sure)

deadalnix commented:
Few nits and this gets in.

deadalnix commented on tests/runner.d:
```diff
@@ -46,6 +46,38 @@ int getInt(string s)
+/* This function is a temporary test for Asm/LLVM  output.
+ * To be removed when the LLVM IR codegen testsuite is in place. */
+bool testAssemblyOutput(string filename, string compiler, bool displayOnlyFailed)
+{
```
egyptian braces

deadalnix commented on tests/runner.d:
```diff
@@ -46,6 +46,38 @@ int getInt(string s)
+ * To be removed when the LLVM IR codegen testsuite is in place. */
+bool testAssemblyOutput(string filename, string compiler, bool displayOnlyFailed)
+{
+    if (compiler == SDC) {
```
tab indent

deadalnix commented on tests/runner.d:
```diff
@@ -46,6 +46,38 @@ int getInt(string s)
+        string[] flagList = [ "-S",
+                              "-S --emit-llvm",
+                              "-c --emit-llvm"
+                            ];
```
string[] flagList = [
      "-S",
      "-S --emit-llvm",
      "-c --emit-llvm",
  ];

JohanEngelen commented:
rebased onto master
